### ISSUE TITLE
Future of courses and Pretendo

### ISSUE THREAD

**Issue by Qiangong2** [RedDuckss/smm1-course-archive#2]
Again, I'm posting this here since the official Pretendo repo doesn't have an issues page.

Over the past two years, I've steadily amassed roughly 4 million courses. As the time when the official SMM servers will be shut down draws closer, I'd like to reach out again and see if there's potential to add these courses to Pretendo. I've brought this up in the discord previously, but haven't gotten much of a response.

So far, it is roughly 140GB uncompressed. I can always continue archiving and throw them up on archive.org when the servers finally go down (or I obtain every single course), but it almost feels like a waste if they can't be contributed to a project like Pretendo.

**Comment by jonbarrow** [comment]
Hi there! First of all, let me say I'm quite impressed you managed to save that many courses! That's really amazing, I'm glad my script was able to help preserve so much data

That being said, I'm unsure if we can actually make use of those courses ourselves. I have brought this up internally with the core team and we have some reservations about rehosting other peoples work like this since attribution would be difficult. Basically every user who's course you have does not have an account with us, and so it would be nearly impossible to actually link all these millions of courses back to their original creator

The solution to that would, of course, be to upload all 4 million+ courses under some "archival" user ran by Pretendo, but that takes away all credit from the original maker

**Comment by Qiangong2** [comment]
Would an option for attribution be creating essentially dummy users? Prepend the username with "nnid_" or something that is restricted on the Pretendo side as far as usernames go? You could do a simple loop to go through and change all of the course metadata nnids. Since they wouldn't be universal users, just SMM specific, they wouldn't necessarily need all the same attributes (and space) that full pretendo users would need

**Comment by jonbarrow** [comment]
It's unclear if anything like that would be feasible. Given usernames have a character limit, there could be issues if prefixing a users username like that would cause issues in the game. For example if a user already has a 16 character username, adding `nnid_` to it will push that username to 21 characters. Also there's nothing stopping people from also registering legitimate usernames prefixed with `nnid_`, and it wouldn't be great for us to implement arbitrary restrictions like that

Also usernames play little role in how things operate under the hood. All titles on the WiiU, 3DS, and Switch rely on an identifier called a PID when referencing users and their data, "usernames" are less identifiers and more just additional metadata used for display purposes as far as the titles are concerned. Given this, it's not reliable to just make "dummy" users since they don't have PIDs. If we randomly assign them PIDs, then that can cause conflicts with legitimate users. For instance, if we assign a "dummy" user the PID of 4000, then in 2 months someone registers a legitimate PNID and also gets assigned the PID 4000, then that becomes a conflict. There's no way to separate the 2 for the game either. We can always store these "dummy" users with the conflicting PID in our database but have them flagged as "dummy" or "nnid" or something, but all the game cares about is the single PID. The moment we hand that conflicting PID to the game, it treats it like a regular user and we lose all context about whether or not it's a "dummy" user or not. It's also not reasonable to allocate a certain pool of millions of PIDs for this since at that point we're just making real PNIDs for these dummy users

Also we wouldn't be able to keep the course IDs the same at all, since the course IDs jump around a lot and our database uses an auto-incrementing counter for IDs. These two systems are not compatible

**Comment by jonbarrow** [comment]
I've brought this up again with the core team to spitball a few ideas on how to handle this situation, and I just wanted to confirm:

You have these courses from using this script correct? https://github.com/PretendoNetwork/smm1-course-archive

Which means that you have both the course data and all the associated metadata JSON?

**Comment by Qiangong2** [comment]
Correct, I've been using that script. It looks like it was updated in February, but I haven't pulled in that change. I'm still using it from 2021

**Comment by jonbarrow** [comment]
That's fine. The updates from a few months ago only add in 3 new pieces of non-essential data. It adds in support for 2 pieces of unknown data and then also how many Miiverse comments a course has, none of which are "essential" here (in Pretendo's implementation these are stubbed anyway)

After talking with the team some more we seem to have come to an agreement on how to use these courses. I cannot disclose how, at this time, since it relates to future in-development plans not yet announced to the public, however it does seem like we have use for these courses after all. Please keep in touch with us either here or on Discord and let us know when you have a complete set!

**Comment by Qiangong2** [comment]
That's great! Is there a way to discern the level code from the ID number? In #1 You mentioned the last available course. However, the metadata only lists the course ID, and not the level code

**Comment by jonbarrow** [comment]
Level codes and course IDs are the same thing. The server only sends the course ID (called the `DataID` internally), and the game constructs the level code you see and share. The level code is made up of 2 parts:

- Checksum
- Course ID (hex encoded)

Though every official level code is more commonly seen as:

- Checksum
- 2 null bytes (`0000`)
- Course ID (hex encoded)

On the server course IDs are sent as uint64's, however the game is incapable of handling IDs this high since the first 2 bytes of the level code is reserved for the checksum. This essentially makes course IDs 48 bit integers instead of 64 bit. On Pretendo we truncate this even further to only a uint32 to keep number storage simple. This means the highest course ID is technically `0xFFFFFFFFFFFF`, or 281,474,976,710,656. Nintendo never exceeded the 32 bit limit though, which is why all official level codes have those null bytes. If you provide the game with a course ID higher than 32 bits then it does use that space in the level code

The checksum bytes use the same algorithm used for PRUDPv0 (one of the transport protocols used by NEX, the software Nintendo uses for multiplayer on the WiiU and 3DS). The checksum is the 3rd and 4th bytes of the md5 hmac of the course ID in little endian, with the hmac key being the md5 hash of the servers access key `9f2b4678`

You mentioned #1, so here's an example of the algorithm using one of the level codes from that issue

```js
const crypto = require('crypto');

const code = 'E97E-0000-0425-6405';
const id = parseInt(code.replace(/-/g, '').substring(4), 16); // * 0x000004256405
const accessKey = '9f2b4678';
const hmacKey = crypto.createHash('md5').update(accessKey).digest(); // * 9ce21c9e046e85f0cf6ca00d1eaaaf5f
const hmacData = Buffer.alloc(0x8);

hmacData.writeBigInt64LE(BigInt(id)); // * 0564250400000000

const hmac = crypto.createHmac('md5', hmacKey).update(hmacData).digest(); // * 2df2e97e9f06b9b761c38107821cc64a
const checksum = hmac.subarray(2, 4).swap16().toString('hex').toUpperCase();

console.log(checksum); // * E97E
```

This checksum is added to try and prevent brute forced spam searches in the game (such as just looking up level code 1, then then level code 2, etc). So you have to also supply the checksum. It's not known why Nintendo decided to make the checksum part of the allocated space for the level ID however, though they likely realized that they would never actually reach 281 trillion uploads so it doesn't matter anyway

If Super Mario Maker ever ever did reach the max course ID though, it's level code would be 2087-FFFF-FFFF-FFFF

**Comment by Qiangong2** [comment]
If you want to start trying to integrate the courses into pretendo, here's 1GB worth to get you started: https://drive.google.com/file/d/1R_kngmA7ApaKHuIIlT-kWAsRlBwtlYnV/view?usp=drive_link

It's courses between 4500111-4585754. I'm currently up to 13834597 and still going. I just had the 45x archive on a different server that I've decommed, so thought I may as well share it here so that it doesn't get lost. 

I also have another 40GB archive of courses, and the current one I'm working on, which is at 94GB. Those will get uploaded later (the current one, once I reach the top end of courses).

**Comment by jneen** [comment]
The way we've done this in Open Course World has been to allow users, having verified that they "own" a particular user account, to manually import their courses from the backup db. This way we can both have the archive and know that the courses we're serving are properly attributed to the creator, and that we have the creator's permission to host them.

**Comment by jneen** [comment]
I've also been working on extracting this project and putting it in a more usable form: https://archive.org/details/super_mario_maker_courses_202105

It claims to be comprehensive but I think it would be a good idea to attempt to consolidate it and your archive once it's done. Yours will definitely have more up-to-date metadata, and it's possible there are courses missing from either (I have two course ids specifically that I've found missing from the archive.org dataset - they seem to have cached a 403.)

**Comment by jonbarrow** [comment]
We have internally discussed how to go about rehosting the courses from this backup and we have already settled on a solution, so we've got that covered

**Comment by jneen** [comment]
Okay! I think it's still important to extract and verify this data, so I will continue to do that.

**Comment by Qiangong2** [comment]
Just to give an update, I'm up to 16623316 and still going. At 118GB at the moment.

**Comment by jneen** [comment]
@Qiangong2 are you applying any sharding to this data? I majorly overloaded my local zfs (to the point where i couldn't `du -sh` the folder anymore) with the sheer number of inodes when attempting to extract the archive.org dataset naively 😅 .

The dataset on archive.org is around 500GB (ASH0 compressed with WARC metadata) and I expect to save about 10% by using squashfs with zstd instead (split into 1000 shards).

**Comment by Qiangong2** [comment]
Hi @jneen! No sharding, I'm not using that dataset at all. I'm using the python script in this repo as it puts it into a format that Pretendo can easily use (since Jon/Pretendo made the script in the first place).

I don't really see the point of sharding when ultimately, it's not that much data to host. As long as it's below a terabyte, it's not a big issue.

**Comment by jneen** [comment]
Okay! Out of curiosity, @Qiangong2 do you have course data for these two ids:

```
63933982
69352261
```

They are in the metadata for the archive.org dataset, but missing in the actual data chunks. The total number of courses in the archive.org dataset is over 10 million, and it's sharded into 1000 chunks. So far in the extraction process I've kept the chunks as they are but it may be that a single squashfs archive can manage them all.

```
$ sqlite3 courses.db 'select count(*) from courses;'
10610412
```

**Comment by Qiangong2** [comment]
The way the script works is by iterating through each level ID, one by one. 

For instance, I'm at 17079693, which is 124GB uncompressed (add my other two archives and we sit at roughly 190GB uncompressed).

If 63933982 is indeed the level ID and not some other id, then that means I have a long way to go...

**Comment by jneen** [comment]
The archive.org dataset is around 500gb "uncompressed" - though it is compressed with ASH0, which your dataset is as well. *Truly* uncompressed it is enough that my extraction process broke at first and I had to split it back into chunks. I have derived from that detaset about 450G of zstd-compressed squashfs files at the moment, and I'm doing some experiments to see if it can be combined into one. If not then I can just ship my sqlite database which associates every level id with a chunk.

**Comment by jonbarrow** [comment]
You can always just modify this script to get the info you want for the courses you're talking about, btw

**Comment by jneen** [comment]
True! I don't have my Nintendo client credentials set up though, I might do that after I finish putting the archive.org dataset into a more usable format.

**Comment by jonbarrow** [comment]
@Qiangong2 @jneen Heads up to both of you, I'm archiving this repository now. All work towards Super Mario Maker archival has been moved to https://github.com/PretendoNetwork/archival-tools

This repository also has a new, updated, archiver which should be much faster (it does not test every single possible course ID one by one), and it downloads ALL data for EVERYTHING, creating a much more accurate backup (at the cost of more storage)
